### DISCUSSION TITLE
Spring WebFlux: How to autowire a service/component in a CustomDirective behavior?

### DISCUSSION THREAD

**Discussion by therealansh** [ExpediaGroup/graphql-kotlin#1914] [Q&A]
I want to autowire a service in a custom directive behaviour. For now I have made a new object only and passed it to the custom wiring. A code snippet here:

```
 @Bean
    fun wiringFactory() = KotlinDirectiveWiringFactory(
        manualWiring = mapOf<String, KotlinSchemaDirectiveWiring>(
            "hasRole" to CustomDirectiveWiring(
                SomeService(APIService(apiBean()))
            )
        )
    )
```
Now this becomes a problem when there are multiple services which need to be autowired.

**Comment by dariuszkuc** [comment]
Hello 👋 
If you are using annotation based configurations you could do

```kotlin
@Component
class SomeService(private val apiService: APIService) {....}

@Component
class APIService(private val apiBean: APIBean) { ... }

// your configuration class then just references the bean
@Configuration
class MyConfiguration {
  // you could also declare wiring as a bean and wire it up directly
  @Bean
  fun wiringFactory(someService: SomeService) = KotlinDirectiveWiringFactory(
    manualWiring = mapOf<String, KotlinSchemaDirectiveWiring>(
      "hasRole" to CustomDirectiveWiring(someService)
    )
  )
}
```

or something along those lines but you still need to define your beans somehow (either through annotations or explicit bean configuration).

**Reply by therealansh** [reply]
hey @dariuszkuc thanks for the reply, however the problem here i see is creating service/components classes thru constructors. For example in your above code, since we need to override the wiring factory in the hooks, we would still need to call this wiringFactory(SomeService()) -> SomeService would be a constructor again right?

**Reply by dariuszkuc** [reply]
Check Spring docs for details on how to autowiring beans works. Yes you could create everything through constructors but you could also rely on autowiring logic to pass those bean references for you.

`@Component class SomeService(private val apiService: APIService) {....}` -> this creates `SomeService` bean (you could also define it using `@Bean` in `@Configuration` class

`@Bean fun wiringFactory(someService: Service)` -> this creates a `wiringFactory` bean that accepts `someService: Service` bean that will be automatically injected by Spring

You would then just inject the `wiringFactory` bean to your custom hooks either through constructor or explicit configuration, i.e.

`@Component class MyCustomHooks(wiringFactory: KotlinDirectiveWiringFactory) { ... }`
vs
`@Bean fun hooks(wiringFactory: KotlinDirectiveWiringFactory) = MyCustomHooks(wiringFactory)`

In either case Spring will automatically inject those beans.
